jdaviz/configs/default/plugins/export_plot/export_plot.py

```diff
@@ -74,3 +112,184 @@ def vue_save_figure(self, filetype):
         the bqplot.Figure save methods.
         """
         self.save_figure(filetype=filetype)
+
+    def _save_movie(self, i_start, i_end, fps, filename, rm_temp_files):
+        # NOTE: All the stuff here has to be in the same thread but
+        #       separate from main app thread to work.
+
+        viewer = self.viewer.selected_obj
+        slice_plg = self.app._jdaviz_helper.plugins["Slice"]._obj
+        orig_slice = slice_plg.slice
+        temp_png_files = []
+        i = i_start
+        video = None
+
+        # TODO: Expose to users?
+        i_step = 1  # Need n_frames check if we allow tweaking
+
+        try:
+            self.movie_recording = True
+
+            while i <= i_end:
+                if self.movie_kill:
+                    break
+
+                slice_plg._on_slider_updated({'new': i})
+                cur_pngfile = f"._cubeviz_movie_frame_{i}.png"
+                self.save_figure(filename=cur_pngfile, filetype="png")
+                temp_png_files.append(cur_pngfile)
+                i += i_step
+
+                # Wait for the roundtrip to the frontend to complete.
+                while viewer.figure._upload_png_callback is not None:
+                    time.sleep(0.05)
+
+            if not self.movie_kill:
+                # Grab frame size.
+                frame_shape = cv2.imread(temp_png_files[0]).shape
+                frame_size = (frame_shape[1], frame_shape[0])
+
+                video = cv2.VideoWriter(filename, cv2.VideoWriter_fourcc(*'mp4v'), fps, frame_size, True)  # noqa: E501
+                for cur_pngfile in temp_png_files:
+                    video.write(cv2.imread(cur_pngfile))
+        finally:

```

Comment:
do we at least need a catch-all except which resets `self.movie_recording = False`?  If I try to run with the spectrum viewer, it raises a snackbar but all buttons are disabled and killing doesn't do anything since there isn't an active thread.

Reply:
Maybe. I keep forgetting about the spectrum viewer. Even better if we can disable the button when spectrum viewer is selected but I don't know how to do that.

Reply:
Is there a reason we can't just support the spectrum viewer?

But either way, the try-finally currently swallows all errors (not just caused by selecting the spectrum viewer) and results in a broken state.

Reply:
Spectrum viewer does not have bqplot save PNG feature because of this:

https://github.com/spacetelescope/jdaviz/blob/a9f13e9cad6de60e5a96611651bf8207e8dd76a7/jdaviz/configs/specviz/plugins/viewers.py#L9

It comes from glue-jupyter, not bqplot. To support it, I think we need to implement stuff over at glue-jupyter. Also, is it really useful to have a movie that just showing the slider slowly... sliding?

Reply:
I added a check to disable the "Export to MP4" button if spectrum-viewer is selected.

> If I try to run with the spectrum viewer, it raises a snackbar but all buttons are disabled and killing doesn't do anything since there isn't an active thread.

It should not disable the buttons at all when you try to record with spectrum viewer, as the exception would be raised before the "movie_recording" flag is even set in the thread because like you say, no thread would be generated, so I have no idea how you got to that state in the first place. But no matter, with the check above, this should not be possible anymore from the GUI.